### ISSUE TITLE
`model_kwargs_cd` is always a copy from `model_kwargs`, is it a bug?

### ISSUE THREAD

**Issue by teamwong111** [DAMO-NLP-SG/VCD#13]
Thanks for your awesome work.

I find `model_kwargs_cd` is always a copy from `model_kwargs` in `while loop` of `vcd_sample` file. In this case, the `past_key_values` of normal input and `past_key_values` of distorted input are same. Is it a bug?

**Comment by teamwong111** [comment]
Two pieces of code with logical conflict

```
model_kwargs_cd = model_kwargs.copy()
```

```
## cd_comments: update model_kwargs_cd for contrastive decoding
if use_cd:
    model_kwargs_cd = self._update_model_kwargs_for_generation(
        outputs_cd, model_kwargs_cd, is_encoder_decoder=self.config.is_encoder_decoder
    )
```

@lixin4ever @hangzhang-nlp  @LengSicong

**Comment by LengSicong** [comment]
Hi, thanks for your interest in our work. 

For your concern, please refer to this [issue](https://github.com/DAMO-NLP-SG/VCD/issues/7).

**Comment by teamwong111** [comment]
#7 comments `I think it does not matter. you can find the process in llava_arch.py -> prepare_inputs_labels_for_multimodal`. It does not resolve my doubts.
1. The `past_key_values` passed in `forward` function are same in normal input case and distorted input case.
2. I want to use vcd in Qwen-vl. I have checked the related files, but there is not similar process likes `prepare_inputs_labels_for_multimodal`.

**Comment by darkpromise98** [comment]
> #7 comments `I think it does not matter. you can find the process in llava_arch.py -> prepare_inputs_labels_for_multimodal`. It does not resolve my doubts.
> 
> 1. The `past_key_values` passed in `forward` function are same in normal input case and distorted input case.
> 2. I want to use vcd in Qwen-vl. I have checked the related files, but there is not similar process likes `prepare_inputs_labels_for_multimodal`.

I have the same question, I found the `next_token_logits` and `next_token_logits_cd` will be equal when the `past_key_values` are same in forward function (That's easy to understand, the `input_ids` is also same)

In this way, the `diffs = (1+cd_alpha)*next_token_logits - cd_alpha*next_token_logits_cd` will be invalid.

**Comment by darkpromise98** [comment]
> #7 comments `I think it does not matter. you can find the process in llava_arch.py -> prepare_inputs_labels_for_multimodal`. It does not resolve my doubts.
> 
> 1. The `past_key_values` passed in `forward` function are same in normal input case and distorted input case.
> 2. I want to use vcd in Qwen-vl. I have checked the related files, but there is not similar process likes `prepare_inputs_labels_for_multimodal`.

It seems that the function `prepare_inputs_labels_for_multimodal ` will utilize the `past_key_values` to compute the logits when `past_key_values` exists and 'input_ids' is single, hence I think the `past_key_values` should not be equal between normal input case and distorted input case.
https://github.com/DAMO-NLP-SG/VCD/blob/master/experiments/llava/model/llava_arch.py#L94

**Comment by SDaoer** [comment]
> > #7 comments `I think it does not matter. you can find the process in llava_arch.py -> prepare_inputs_labels_for_multimodal`. It does not resolve my doubts.
> > 
> > 1. The `past_key_values` passed in `forward` function are same in normal input case and distorted input case.
> > 2. I want to use vcd in Qwen-vl. I have checked the related files, but there is not similar process likes `prepare_inputs_labels_for_multimodal`.
> 
> I have the same question, I found the `next_token_logits` and `next_token_logits_cd` will be equal when the `past_key_values` are same in forward function (That's easy to understand, the `input_ids` is also same)
> 
> In this way, the `diffs = (1+cd_alpha)*next_token_logits - cd_alpha*next_token_logits_cd` will be invalid.

same confusion, it seems that the distorted image will only be used for the generation of first predicted token:

```
        # if `inputs_embeds` are passed, we only want to use them in the 1st generation step
        if inputs_embeds is not None and past_key_values is None:
            model_inputs = {"inputs_embeds": kwargs.get("images_cd")}
        else:
            model_inputs = {"input_ids": input_ids}
```

**Comment by darkpromise98** [comment]
> > > #7 comments `I think it does not matter. you can find the process in llava_arch.py -> prepare_inputs_labels_for_multimodal`. It does not resolve my doubts.
> > > 
> > > 1. The `past_key_values` passed in `forward` function are same in normal input case and distorted input case.
> > > 2. I want to use vcd in Qwen-vl. I have checked the related files, but there is not similar process likes `prepare_inputs_labels_for_multimodal`.
> > 
> > 
> > I have the same question, I found the `next_token_logits` and `next_token_logits_cd` will be equal when the `past_key_values` are same in forward function (That's easy to understand, the `input_ids` is also same)
> > In this way, the `diffs = (1+cd_alpha)*next_token_logits - cd_alpha*next_token_logits_cd` will be invalid.
> 
> same confusion, it seems that the distorted image will only be used for the generation of first predicted token:
> 
> ```
>         # if `inputs_embeds` are passed, we only want to use them in the 1st generation step
>         if inputs_embeds is not None and past_key_values is None:
>             model_inputs = {"inputs_embeds": kwargs.get("images_cd")}
>         else:
>             model_inputs = {"input_ids": input_ids}
> ```

This problem can be fixed by ignoring the copy operation, and the experimental results will not be affected in POPE evaluation, since the output token is single ('Yes'or 'No').
(https://github.com/yfzhang114/LLaVA-Align/blob/main/vcd_utils/vcd_sample.py#L91)(https://github.com/yfzhang114/LLaVA-Align/blob/main/vcd_utils/vcd_sample.py#L152)

**Comment by Stevetich** [comment]
> > > > #7 comments `I think it does not matter. you can find the process in llava_arch.py -> prepare_inputs_labels_for_multimodal`. It does not resolve my doubts.
> > > > 
> > > > 1. The `past_key_values` passed in `forward` function are same in normal input case and distorted input case.
> > > > 2. I want to use vcd in Qwen-vl. I have checked the related files, but there is not similar process likes `prepare_inputs_labels_for_multimodal`.
> > > 
> > > 
> > > I have the same question, I found the `next_token_logits` and `next_token_logits_cd` will be equal when the `past_key_values` are same in forward function (That's easy to understand, the `input_ids` is also same)
> > > In this way, the `diffs = (1+cd_alpha)*next_token_logits - cd_alpha*next_token_logits_cd` will be invalid.
> > 
> > 
> > same confusion, it seems that the distorted image will only be used for the generation of first predicted token:
> > ```
> >         # if `inputs_embeds` are passed, we only want to use them in the 1st generation step
> >         if inputs_embeds is not None and past_key_values is None:
> >             model_inputs = {"inputs_embeds": kwargs.get("images_cd")}
> >         else:
> >             model_inputs = {"input_ids": input_ids}
> > ```
> 
> This problem can be fixed by ignoring the copy operation, and the experimental results will not be affected in POPE evaluation, since the output token is single ('Yes'or 'No'). (https://github.com/yfzhang114/LLaVA-Align/blob/main/vcd_utils/vcd_sample.py#L91)(https://github.com/yfzhang114/LLaVA-Align/blob/main/vcd_utils/vcd_sample.py#L152)

So the copy operation is not correct and needs to be deleted to make the contrastive decoding work for the whole sequence, right?

**Comment by Stevetich** [comment]
Actually, I have mentioned the problem of the same past_key_values in #7 and it has confused me for a really long time... So can I understand in this way that `model_kwargs` and `model_kwargs_cd` need to save the `past_key_values` separately, right?

**Comment by darkpromise98** [comment]
> > > > > #7 comments `I think it does not matter. you can find the process in llava_arch.py -> prepare_inputs_labels_for_multimodal`. It does not resolve my doubts.
> > > > > 
> > > > > 1. The `past_key_values` passed in `forward` function are same in normal input case and distorted input case.
> > > > > 2. I want to use vcd in Qwen-vl. I have checked the related files, but there is not similar process likes `prepare_inputs_labels_for_multimodal`.
> > > > 
> > > > 
> > > > I have the same question, I found the `next_token_logits` and `next_token_logits_cd` will be equal when the `past_key_values` are same in forward function (That's easy to understand, the `input_ids` is also same)
> > > > In this way, the `diffs = (1+cd_alpha)*next_token_logits - cd_alpha*next_token_logits_cd` will be invalid.
> > > 
> > > 
> > > same confusion, it seems that the distorted image will only be used for the generation of first predicted token:
> > > ```
> > >         # if `inputs_embeds` are passed, we only want to use them in the 1st generation step
> > >         if inputs_embeds is not None and past_key_values is None:
> > >             model_inputs = {"inputs_embeds": kwargs.get("images_cd")}
> > >         else:
> > >             model_inputs = {"input_ids": input_ids}
> > > ```
> > 
> > 
> > This problem can be fixed by ignoring the copy operation, and the experimental results will not be affected in POPE evaluation, since the output token is single ('Yes'or 'No'). (https://github.com/yfzhang114/LLaVA-Align/blob/main/vcd_utils/vcd_sample.py#L91)(https://github.com/yfzhang114/LLaVA-Align/blob/main/vcd_utils/vcd_sample.py#L152)
> 
> So the copy operation is not correct and needs to be deleted to make the contrastive decoding work for the whole sequence, right?

I think so. However, this is my personal opinion, it would be best if the author could respond to this question.

**Comment by LengSicong** [comment]
> > > > #7 comments `I think it does not matter. you can find the process in llava_arch.py -> prepare_inputs_labels_for_multimodal`. It does not resolve my doubts.
> > > > 
> > > > 1. The `past_key_values` passed in `forward` function are same in normal input case and distorted input case.
> > > > 2. I want to use vcd in Qwen-vl. I have checked the related files, but there is not similar process likes `prepare_inputs_labels_for_multimodal`.
> > > 
> > > 
> > > I have the same question, I found the `next_token_logits` and `next_token_logits_cd` will be equal when the `past_key_values` are same in forward function (That's easy to understand, the `input_ids` is also same)
> > > In this way, the `diffs = (1+cd_alpha)*next_token_logits - cd_alpha*next_token_logits_cd` will be invalid.
> > 
> > 
> > same confusion, it seems that the distorted image will only be used for the generation of first predicted token:
> > ```
> >         # if `inputs_embeds` are passed, we only want to use them in the 1st generation step
> >         if inputs_embeds is not None and past_key_values is None:
> >             model_inputs = {"inputs_embeds": kwargs.get("images_cd")}
> >         else:
> >             model_inputs = {"input_ids": input_ids}
> > ```
> 
> This problem can be fixed by ignoring the copy operation, and the experimental results will not be affected in POPE evaluation, since the output token is single ('Yes'or 'No'). (https://github.com/yfzhang114/LLaVA-Align/blob/main/vcd_utils/vcd_sample.py#L91)(https://github.com/yfzhang114/LLaVA-Align/blob/main/vcd_utils/vcd_sample.py#L152)

Hi all, thanks a lot for your interest in our work and active discussions!

We are sorry that we made this mistake when cleaning the open-source code. 

To make it clear:
1. The VCD mechanisms require two sets of past_key_values during the inference, one for the original image input and the other for the distorted image.
2. As explained by @darkpromise98 (thanks!:)), this problem will not affect the POPE result since there is only one output token.
3. We made this mistake only for the open-sourced cleaned codes. The experimental results reported in our paper follow the correct logic and are guaranteed to be faithful and reproducible. 
4. We will correct the public codes as soon as possible. Instead of removing the copy operation directly, the copy should be moved before the [while true](https://github.com/yfzhang114/LLaVA-Align/blob/main/vcd_utils/vcd_sample.py#L93) loop only for the generation of the first token.

**Comment by dacian7** [comment]
> 2\. Qwen-vl.

@teamwong111 Hi, have you successful solved this issue? I am also modifying Qwen family models recently and I appreciate it if we could have a discussion about that! Thanks!

**Comment by natsunoshion** [comment]
> > 2. Qwen-vl.
> 
> [@teamwong111](https://github.com/teamwong111) Hi, have you successful solved this issue? I am also modifying Qwen family models recently and I appreciate it if we could have a discussion about that! Thanks!

I think this is a problem. When I try to use it on qwen2.5vl and other models, I find it wrong when apply `model_kwargs = self._get_initial_cache_position(input_ids, model_kwargs)`, which is essential for these models. It will make the `past_key_values` messed up between model_kwargs and model_kwargs_cd.

I solve it by using `deepcopy` instead of normal copy. Like:

```python
model_kwargs_cd = copy.deepcopy(model_kwargs)
```

**Comment by alterego238** [comment]
> > > 2. Qwen-vl.
> > 
> > 
> > [@teamwong111](https://github.com/teamwong111) Hi, have you successful solved this issue? I am also modifying Qwen family models recently and I appreciate it if we could have a discussion about that! Thanks!
> 
> I think this is a problem. When I try to use it on qwen2.5vl and other models, I find it wrong when apply `model_kwargs = self._get_initial_cache_position(input_ids, model_kwargs)`, which is essential for these models. It will make the `past_key_values` messed up between model_kwargs and model_kwargs_cd.
> 
> I solve it by using `deepcopy` instead of normal copy. Like:
> 
> model_kwargs_cd = copy.deepcopy(model_kwargs)

Hello, I'm also trying on qwen2.5vl and find the similar problem. I also try `model_kwargs_cd = copy.deepcopy(model_kwargs)`, but it doesn't work perfectly. The logits are still wrong except the first generated token. I have some other tries, can you share your code, or we can discuss it if possible? My email: jialhon@gmail.com

**Comment by daiqing98** [comment]
> > > > 2. Qwen-vl.
> > > 
> > > 
> > > [@teamwong111](https://github.com/teamwong111) Hi, have you successful solved this issue? I am also modifying Qwen family models recently and I appreciate it if we could have a discussion about that! Thanks!
> > 
> > 
> > I think this is a problem. When I try to use it on qwen2.5vl and other models, I find it wrong when apply `model_kwargs = self._get_initial_cache_position(input_ids, model_kwargs)`, which is essential for these models. It will make the `past_key_values` messed up between model_kwargs and model_kwargs_cd.
> > I solve it by using `deepcopy` instead of normal copy. Like:
> > model_kwargs_cd = copy.deepcopy(model_kwargs)
> 
> Hello, I'm also trying on qwen2.5vl and find the similar problem. I also try `model_kwargs_cd = copy.deepcopy(model_kwargs)`, but it doesn't work perfectly. The logits are still wrong except the first generated token. I have some other tries, can you share your code, or we can discuss it if possible? My email: [jialhon@gmail.com](mailto:jialhon@gmail.com)

@alterego238 Hi, thanks for your discussion. I guess qwen2 does not have this issue, right?

**Comment by alterego238** [comment]
> > > > > 2. Qwen-vl.
> > > > 
> > > > 
> > > > [@teamwong111](https://github.com/teamwong111) Hi, have you successful solved this issue? I am also modifying Qwen family models recently and I appreciate it if we could have a discussion about that! Thanks!
> > > 
> > > 
> > > I think this is a problem. When I try to use it on qwen2.5vl and other models, I find it wrong when apply `model_kwargs = self._get_initial_cache_position(input_ids, model_kwargs)`, which is essential for these models. It will make the `past_key_values` messed up between model_kwargs and model_kwargs_cd.
> > > I solve it by using `deepcopy` instead of normal copy. Like:
> > > model_kwargs_cd = copy.deepcopy(model_kwargs)
> > 
> > 
> > Hello, I'm also trying on qwen2.5vl and find the similar problem. I also try `model_kwargs_cd = copy.deepcopy(model_kwargs)`, but it doesn't work perfectly. The logits are still wrong except the first generated token. I have some other tries, can you share your code, or we can discuss it if possible? My email: [jialhon@gmail.com](mailto:jialhon@gmail.com)
> 
> [@alterego238](https://github.com/alterego238) Hi, thanks for your discussion. I guess qwen2 does not have this issue, right?

@daiqing98 Sorry, I haven't tried qwen2vl yet.

**Comment by rickeyhhh** [comment]
Have you solved this issue? I meet this on llava-next too.
